memtable, cache: Eagerly compact data with tombstones

tgrabiec opened the pull request:
When memtable receives a tombstone it can happen under some workloads
that it covers data which is still in the memtable. Some workloads may
insert and delete data within a short time frame. We could reduce the
rate of memtable flushes if we eagerly drop tombstoned data.

One workload which benefits is the raft log. It stores a row for each
uncommitted raft entry. When entries are committed they are
deleted. So the live set is expected to be short under normal
conditions.

Fixes #652.

avikivity commented on partition_version.cc:
```diff
@@ -326,24 +326,33 @@ partition_version& partition_entry::add_version(const schema& s, cache_tracker*
         mp.upgrade(mp_schema, s);
     }
     auto new_version = current_allocator().construct<partition_version>(std::move(mp));
+    partition_snapshot_ptr snp; // Should die after new_version is inserted
```
Will this lose performance?

tgrabiec replied:
No, it's only used when preempted.

avikivity commented on partition_version.cc:
```diff
@@ -176,27 +176,38 @@ stop_iteration partition_snapshot::merge_partition_versions(mutation_application
         // This is good for performance because in case we were at the latest version
         // we leave it for incoming writes and they don't have to create a new one.
         partition_version* current = &*v;
+        int version_no = 0;
```
version_number_type?

avikivity commented on mutation_partition.cc:
```diff
@@ -305,14 +305,104 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
+            }
+
+            if (need_preempt() && i != end) {
+                res = apply_resume(stage, position_in_partition(i->position()));
```
What happens if copying causes us to throw here?

I'm guessing we'll be left in some undefined state. Maybe better to swallow the exception and not stop in that case.

avikivity commented on mutation_partition.cc:
```diff
@@ -305,14 +305,104 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
+        return stop_iteration::yes;
+    };
+
+    if (res._stage <= apply_resume::stage::range_tombstone_compaction) {
```
This really reminds me of a coroutine. We can consider switching to real coroutines here, though I wouldn't touch it now.

avikivity replied:
(real coroutines, but not seastar coroutines)

avikivity commented on mutation_partition.cc:
```diff
@@ -305,14 +305,104 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
+            return stop_iteration::no;
+        }
+        // TODO: Drop redundant range tombstones
+        p._tombstone = {};
```
aha

avikivity commented on test/boost/memtable_test.cc:
```diff
@@ -572,6 +573,74 @@ SEASTAR_THREAD_TEST_CASE(test_tombstone_compaction_during_flush) {
+        ss.add_row(m, ss.make_ckey(2), "v1");
+        ss.add_row(m, ss.make_ckey(3), "v1");
+        ss.add_row(m, ss.make_ckey(4), "v1");
+        mt->apply(m);
```
Shouldn't it be much larger to force preemption?

tgrabiec replied:
Not in debug mode.

avikivity commented:
Looks good, some comments.

avikivity commented on replica/table.cc:
```diff
@@ -776,6 +776,8 @@ void table::set_metrics() {
                 ms::make_counter("memtable_row_writes", _stats.memtable_app_stats.row_writes, ms::description("Number of row writes performed in memtables"))(cf)(ks),
                 ms::make_counter("memtable_row_hits", _stats.memtable_app_stats.row_hits, ms::description("Number of rows overwritten by write operations in memtables"))(cf)(ks),
+                ms::make_counter("memtable_rows_dropped_by_tombstones", _stats.memtable_app_stats.rows_dropped_by_tombstones, ms::description("Number of rows scanned during write of a tombstone for the purpose of compaction in memtables"))(cf)(ks),
+                ms::make_counter("memtable_rows_compacted_with_tombstones", _stats.memtable_app_stats.rows_compacted_with_tombstones, ms::description("Number of rows dropped in memtables by a tombstone write"))(cf)(ks),
```
I'd like the same counters for cache, when a delete misses the memtable entry but meets it in cache.

tgrabiec replied:
Ok

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/445/

avikivity commented on mutation_partition.cc:
```diff
@@ -305,14 +305,104 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
+    // Compacts rows in [i, end) with the tombstone.
+    // Erases entries which are left empty by compaction.
+    // Does not affect continuity.
+    auto apply_tombstone_to_rows = [&] (apply_resume::stage stage, tombstone tomb, rows_type::iterator i, rows_type::iterator end) -> stop_iteration {
```
Can we solve #2577 in the same way? Instead of apply_tombstones_to_rows we apply_schema_upgrade_to_rows, but otherwise same.

tgrabiec replied:
I think it's simpler not to. We can maintain the invariant that each version conforms to a single schema. Then upgrading schema will have to be done by the means of moving atoms from one version to another. The only change we need to do here is to teach the algorithm to upgrade schema of the moved elements.

bhalevy commented on mutation_partition.cc:
```diff
@@ -305,14 +305,106 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
+
+            if (all_dead && e.row().empty() && inside_continuous_range) {
+                ++app_stats.rows_dropped_by_tombstones;
+                i = _rows.erase(i);
```
Shouldn't we get a reference to `i` instead of a copy by value so if we preempt we'll continue from where we returned last time?

bhalevy commented on mutation_partition.cc:
```diff
@@ -305,14 +305,106 @@ stop_iteration mutation_partition::apply_monotonically(const schema& s, mutation
+            tomb.apply(rt.tombstone().tomb);
+
+            if (apply_tombstone_to_rows(apply_resume::stage::range_tombstone_compaction, tomb, i, end) == stop_iteration::no) {
+                return stop_iteration::no;
```
where do we set `res._pos` if we preemept?

bhalevy replied:
ok, I see. `res = apply_resume` above...

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/563/

bhalevy commented:
@tgrabiec please see failures in CI.

tgrabiec commented:
In V3:

  - Use `version_number_type`

  - Allocate apply_resume contents inside standard allocator (fixes alloc-dealloc mismatch if snapshot dies mid-merge)

  - Add metrics for compaction happening in the row cache

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/579/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/597/

tgrabiec commented:
Looks like a flaky test:

```
memtable_test: test/boost/memtable_test.cc:920: auto failed_flush_prevents_writes::run_test_case()::(anonymous class)::operator()(cql_test_env &) const: Assertion `eventually_true([&db]() { return db.cf_stats()->failed_memtables_flushes_count != 0; })' failed.
```

\cc @mikolajsieluzycki

mikolajsieluzycki commented:
```
./test.py --repeat 1000 boost/memtable_test --mode=dev
Found 1000 tests.
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[1000/1000]  boost     dev   [ PASS ] memtable_test.1000           
------------------------------------------------------------------------------
```
The assert suggests that either
- error injection didn't work
- explicitly called soft pressure on dirty memory manager didn't cause flush. This can happen only if
 - the db shutdown was requested (possible)
 - there was another background/explicit flush going on (the test doesn't trigger any)
 - the pressure was released before flush (doesn't happen)
 - the flushed memtable doesn't have any data (it does)

The test report is empty https://jenkins.scylladb.com/job/releng/job/Scylla-CI/579/testReport/ and the only case that could potentially cause the test failure is not flushing because the db shutdown was requested. The test has been in place for 4months and I'd be surprised if it started being flaky now. Could your changes affect any of the mechanisms I mentioned above that would prevent a flush from happening?

bhalevy commented:
@mikolajsieluzycki the test output is kept under the artifacts/testlog folder.
See https://jenkins.scylladb.com/job/releng/job/Scylla-CI/579/artifact/testlog/x86_64_dev/memtable_test.3.log

bhalevy commented:
> Looks like a flaky test:
> 
> ```
> memtable_test: test/boost/memtable_test.cc:920: auto failed_flush_prevents_writes::run_test_case()::(anonymous class)::operator()(cql_test_env &) const: Assertion `eventually_true([&db]() { return db.cf_stats()->failed_memtables_flushes_count != 0; })' failed.
> ```
> 
> \cc @mikolajsieluzycki

https://jenkins.scylladb.com/job/scylla-master/job/next/4921/artifact/testlog/x86_64_dev/memtable_test.8.log

tgrabiec commented:
I'm looking into this

tgrabiec commented:
In V4:

  - fixed flaky (but likely) failure of failed_flush_prevents_writes in dev mode
  
    The test was making some assumptions which not always hold after the change. The fix is in commit https://github.com/scylladb/scylla/pull/10612/commits/8d8f4b76fe0e6b697a896eb6e68a79c1a0ba87f0

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/622/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/649/

tgrabiec commented:
In V5:

  - Fixed  [memtable: Subtract from flushed memory when cleaning](https://github.com/scylladb/scylla/pull/10612/commits/31ae3085f754240448155be8627d8147af389c95) to also account for bytes released during unsuccessful allocating section executions. Fixes sporadic failure in database_test_g in debug mode.

avikivity commented on mutation_cleaner.hh:
```diff
@@ -32,19 +32,23 @@ private:
-            mutation_application_stats& app_stats, seastar::scheduling_group sg = seastar::current_scheduling_group())
+            mutation_application_stats& app_stats,
+            seastar::scheduling_group sg = seastar::current_scheduling_group(),
+            std::function<void(size_t)> on_space_freed = nullptr)
```
Tangential: we may be able to get rid of byte-granularity accounting with compaction groups.

Let's say we have a largish number N (100-1000) of compaction groups. Currently, we measure the memtable fill bandwidth and memtable outgoing bandwidth by using the memtable byte size, and try to match them by increasing/decreasing flush shares. But if we have a large number of compaction groups then the flush bandwidth can be measured by (occupied size of regions of recently flushed memtables) / (time to flush those memtables). This reacts more slowly to changes, but simplifies the code and removes double accounting.

avikivity commented:
Dequeuing, see #10720

bhalevy commented:
@tgrabiec please look into the failures reported on #10720

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/684/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/702/

tgrabiec commented:
In V6:

 - Fixed a bug in compacting reader uncovered by the series, causing random failures in flat_mutation_reader_test: https://github.com/scylladb/scylla/pull/10612/commits/604e720706b01a720abe683e49de8dbcd12a458c
 
 - Rebased, which also fixes a failure in cdc test (fix was committed to master since V5)
